### DISCUSSION TITLE
Pure rust chainboot

### DISCUSSION THREAD

**Discussion by berkus** [rust-embedded/rust-raspberrypi-OS-tutorials#173] [General]
Was trying to find through which medium to contact @andre-richter, but hey, this is as good as any.

Some fun: [pure Rust chain bootloader for Raspberry Pi](https://github.com/metta-systems/vesper/tree/78c3249bf315f7b53a90230cba03259421c0fd20/bin/chainboot)

:)

**Comment by andre-richter** [comment]
Nice! I actually went the exact different route and returned to doing basic init in assembly after reading https://github.com/rust-embedded/cortex-m-rt/issues/300.

This decision was later reinforced when I realized that you cannot rely on the compiler to not reorder register manipulations done through `cortex-a`, aka tock-registers, with other operations. Unfortunately I don’t remember exactly where I observed it in a problematic situation, and more unfortunately, I never followed through documenting it.

Might have been a situation where I wanted to re-set the stack pointer, but function prologue operations always came in first. But naked functions might have helped, so I’m not sure if it was this example.

All in all, I personally feel more safe having bss and stack prepared in assembly, and thus having the basics of the abstract machine sound when we start with Rust.

In your code's comments I saw the reference to https://github.com/rust-embedded/cortex-m-rt/issues/300 as well, but haven’t yet checked the other resources you mentioned regarding this discussion.

**Comment by berkus** [comment]
Yeah, I took note of pointer provenance woes and switched to pointer+size to make it valid.

While messing SP is still a possibility, making the very first function extremely small (just set SP and call follow up function) makes it very unlikely. I will keep an eye on this, but so far happy to make it boot without involving any asm.
